### PR TITLE
Add archiver task to handle orphaned objects

### LINKED ISSUE #440: Orphaned records in tool_objectfs_objects create additional CPU load
On some larger sites the deleter and pusher candidate queries are responsible for quite a bit of load on the database. I've identified a large number of orphaned records in the mdl_tool_objectfs_objects table, i.e. records where there is no longer a matching contenthash in the mdl_files table. These orphaned records are substantially increasing the size of joins in these candidate queries which are run very regularly by cron. Because the records have been orphaned their state doesn't progress, and they are included in the join every time these queries are run by cron.

A mdl_tool_objectfs_object record will be orphaned where a file has been uploaded to Moodle, the file has been copied to external storage, e.g. S3, and then a user or some other process has deleted the file from within Moodle, resulting in the mdl_files record with the matching contenthash being deleted. I've run the following query on one large site which shows close to 2 million of these orphaned records in different states, with the bulk in 1 (duplicated) and 2 (remote only):

```
SELECT o.location, count(*)
FROM mdl_tool_objectfs_objects o
	left join mdl_files f ON o.contenthash = f.contenthash
WHERE f.id is null
	AND o.contenthash != 'da39a3ee5e6b4b0d3255bfef95601890afd80709' --empty files (e.g. ".")
GROUP BY o.location
```

If we were to archive these files by setting location to an "archived" state such as -2, then this would reduce the size of the joins in the deleter and pusher candidate queries.

On my site the deleter candidate query runs for 37 seconds before archiving any orphaned records. After archiving this same query runs in 1 second.

We would need to account for situations where an "archived" file is re-uploaded to moodle at a later date so the location can be updated to 1 (duplicated) or 0 (never copied to external storage) but this already looks to be covered adequately through the various calls to manager::get_object_location_from_hash() in the various add_file*** functions in class object_file_system.

### PR DESCRIPTION
**Pull request by keevan**
Closes #440 

Added reporting support, unit tests, typehints, comments and cleaned up the base implementation.

Screenshots of admin setting:

image

Few screenshots of the reporting:

image

image

### CODE REVIEW

File: classes/task/delete_archived_object_metadata.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    protected $stringname = 'delete_archived_object_metadata_task';
+
+    /**
+     * Execute task
+     */
+    public function execute() {
+        global $DB;
+
+        $wheresql = 'location = :location and timeduplicated < :ageforremoval';
+        $ageforremoval = 30 * DAYSECS;
```
Comment (brendanheywood) [inline classes/task/delete_archived_object_metadata.php:45]: turn this into a setting which defaults to 0 which means don't ever delete

File: db/tasks.php [hunk 2]
```diff
@@ -53,6 +53,24 @@
         'dayofweek' => '*',
         'month'     => '*'
     ),
+    array(
+        'classname' => 'tool_objectfs\task\archive_orphaned_objects',
+        'blocking'  => 0,
+        'minute'    => '52',
+        'hour'      => '*/12',
```
Comment (brendanheywood) [inline db/tasks.php:60]: make this once a day

File: db/tasks.php [hunk 3]
```diff
@@ -53,6 +53,24 @@
...
+        'day'       => '*',
+        'dayofweek' => '*',
+        'month'     => '*'
+    ),
+    array(
+        'classname' => 'tool_objectfs\task\delete_archived_object_metadata',
+        'blocking'  => 0,
+        'minute'    => 'R',
+        'hour'      => '*/12',
```
Comment (brendanheywood) [inline db/tasks.php:69]: make this once a day

File: classes/task/delete_archived_object_metadata.php (hunk 1 above)
Comment (keevan) [reply classes/task/delete_archived_object_metadata.php]: image

I've added this under general settings, please let me know if the wording needs to be updated.

**Comment by keevan** [comment]
Moodle 3.9.11+ (Build: 20211203), 567cf86dad927cde531f9ba9a84fea40cb61d2a9
Php: 7.3.28, pgsql: 10.17 (Debian 10.17-1.pgdg90+1), OS: Linux 5.4.0-91-generic x86_64
PHPUnit 7.5.20 by Sebastian Bergmann and contributors.

...............................................................  63 / 211 ( 29%)
.............................S.SSSSSSSSSSSSSSSSSS.SSSSSSSSSSSSS 126 / 211 ( 59%)
SSSSSSSSS.........SSS.......................................S.. 189 / 211 ( 89%)
......................                                          211 / 211 (100%)

Time: 19.71 seconds, Memory: 636.00 MB

OK, but incomplete, skipped, or risky tests!
Tests: 211, Assertions: 274, Skipped: 45.

**Pull request merged**